### ISSUE TITLE
[ch04] 요약 및 정리

### ISSUE THREAD

**Issue by JaehoNam** [JaehoNam/aas-practice#6]
### 4.1 회귀로 돌아와서

The common thread linking regression and classification is that both involve predicting one (or more) values given one (or more) other values. To do so, both require a body of inputs and outputs to learn from. They need to be fed both questions and known answers. For this reason, they are known as types of supervised learning.(지도 학습의 두 가지 큰 예시 - 회귀, 분류)

This chapter will focus on a popular and flexible type of algorithm for both classification and regression: decision trees, and the algorithm’s extension, random decision forests.(랜덤 포레스트)

### 4.2 벡터와 특징

These five features together, in order, are known as a feature vector, and can describe any day’s weather. This usage bears some resemblance to use of the term vector in linear algebra, except that a vector in this sense can conceptually contain nonnumeric values, and even lack some values. For purposes of discussion, this book will talk about features in two broad groups only: categorical features and numeric features.(특징 벡터, 범주형 특징과 수치형 특징)

### 4.3 학습 예제

It’s not uncommon to simply include the target as another feature in the feature vector.(목표를 입력 특징 벡터에 포함시킴)

Note that regression problems are just those where the target is a numeric feature, and classification problems are those where the target is categorical. Not every regression or classification algorithm can handle categorical features or categorical targets; some are limited to numeric features.(모든 회귀 분류 알고리즘이 범주형 특징을 다룰 수 있는 것은 아님)

### 4.4 의사 결정 나무와 랜덤 포레스트

It turns out that the family of algorithms known as decision trees can naturally handle both categorical and numeric features. Building a single tree can be done in parallel, and many trees can be built in parallel at once. They are robust to outliers in the data, meaning that a few extreme and possibly erroneous data points might not affect predictions at all. They can consume data of different types and on different scales without the need for preprocessing or normalization.(의사 결정 나무에 속하는 알고리즘은 범주형 특징을 다룰 수 있음, 데이터 내의 이상치에 잘 휘둘리지 않음, 전처리 정규화 과정 없이도 다룰 수 있음)

This series of yes/no decisions that lead to a prediction are what decision trees embody. In this sense, it is natural to think of the process as a tree of decisions, where each internal node in the tree is a decision, and each leaf node is a final answer.(의사 결정 트리 -> 일련의 예/아니오 선택)

### 4.5 Covtype 데이터셋

The data set records the types of forest-covering parcels of land in Colorado, USA.(콜로라도 안에 있는 숲 토지 구획 데이터)

The forest cover type is to be predicted from the rest of the features, of which there are 54 in total. There are 581,012 examples in the data set, which does not exactly qualify as big data but is large enough to be manageable as an example and still highlight some issues of scale.(54개의 column, 총 58만 개 데이터)

### 4.6 데이터 준비하기

```
val colNames = Seq(
  "Elevation", "Aspect", "Slope",
  "Horizontal_Distance_To_Hydrology", "Vertical_Distance_To_Hydrology",
  "Horizontal_Distance_To_Roadways",
  "Hillshade_9am", "Hillshade_Noon", "Hillshade_3pm",
  "Horizontal_Distance_To_Fire_Points"
) ++ (
  (0 until 4).map(i => s"Wilderness_Area_$i")
) ++ (
  (0 until 40).map(i => s"Soil_Type_$i")
) ++ Seq("Cover_Type")
```

위와 같은 코드로 column naming

### 4.7 첫 번째 의사 결정 나무

The output doesn’t look exactly like a sequence of numbers, but that’s because this shows a raw representation of the vector, represented as a SparseVector instance to save storage. Because most of the 54 values are 0, it only stores nonzero values and their indices.(VectorAssembler -> 0이 아닌 column 만 특성 벡터로 뽑아줌)

VectorAssembler is an example of Transformer within the current Spark MLlib "Pipelines" API. It transforms another DataFrame into a DataFrame, and is composable with other transformations into a pipeline.(VectorAssembler -> 일종의 transformer)

The resulting DecisionTreeClassificationModel is itself a transformer because it can transform a data frame containing feature vectors into a data frame also containing predictions.(결과도 일종의 transformer)

Based on this snippet, it looks like the model could use some work. Its predictions look like they are often wrong. As with the ALS implementation, the DecisionTreeClassifier implementation has several hyperparameters for which a value must be chosen, and they’ve all been left to defaults here.(현재 hyperparameter tuning 이 없어서 낮은 정확도를 보임)

MulticlassClassificationEvaluator can compute accuracy and other metrics that evaluate the quality of the model’s predictions. It’s an example of an evaluator in Spark MLlib, which is responsible for assessing the quality of an output DataFrame in some way.(MulticlassClassificationEvaluator -> acc, f1 확인 가능 -> 대략 0.7)

Fortunately, Spark provides support code to compute the confusion matrix. Unfortunately, that implementation exists as part of the older MLlib APIs that operate on RDDs. However, that’s no big deal, because data frames and data sets can freely be turned into RDDs and used with these older APIs. Here, MulticlassMetrics is appropriate for a data frame containing predictions.(MulticlassMetrics -> confusion matrix)

Random guessing achieves 37% accuracy then, which makes 70% seem like a good result after all. But this result was achieved with default hyperparameters. We can do even better by exploring what these actually mean for the tree-building process.(찍어서 나오는 확률 37% 보다 70% 는 훨씬 큰 값으로 좋아 보이지만 더 좋은 모델 만들 수 있음)

### 4.8 의사 결정 나무 하이퍼파라미터

Maximum depth simply limits the number of levels in the decision tree. It is the maximum number of chained decisions that the classifier will make to classify an example.(최대 깊이, 오버피팅과 관련)

So, the set of decision rules to try is really a set of values to plug in to the decision rule. A larger number of bins requires more processing time but might lead to finding a more optimal decision rule.(bin 단위, bin 크기가 작을수록 시간은 오래 걸리지만 정확도 향상)

Put another way, good rules divide the training data’s target values into relatively homogeneous, or “pure,” subsets. Picking a best rule means minimizing the impurity of the two subsets it induces. There are two commonly used measures of impurity: Gini impurity and entropy.(여러 class 중 하나의 class 를 특정할수록, 즉 impurity 가 0에 가까울수록 좋은 규칙 -> impurity 측정 방식에는 대표적으로 Gini impurity 와 entropy 가 있음)

- Gini impurity
  <img width="171" alt="Screen Shot 2021-03-08 at 3 51 01 AM" src="https://user-images.githubusercontent.com/32262002/110250952-81524400-7fc1-11eb-9306-4c91c2bcd3a9.png">
- entropy
  <img width="318" alt="Screen Shot 2021-03-08 at 3 51 27 AM" src="https://user-images.githubusercontent.com/32262002/110250962-9202ba00-7fc1-11eb-98e3-91eed6f77b6e.png">

One or the other measure may be a better metric for picking decision rules in a given data set. They are, in a way, similar. Both involve a weighted average. The default in Spark’s implementation is Gini impurity.(둘 다 일종의 가중 평균 개념이 들어가 있음, 스파크 기본은 Gini impurity)

Finally, minimum information gain is a hyperparameter that imposes a minimum information gain, or decrease in impurity, for candidate decision rules. Rules that do not improve the subsets impurity enough are rejected.(minimum information gain, impurity 를 일정 이상 감소시키지 못하면 기각, 오버피팅 방지)

### 4.9 의사 결정 나무 튜닝하기

```
val pipeline = new Pipeline().setStages(Array(assembler, classifier))
```
먼저 위의 코드로 pipeline 을 만든다. 또한, `ParamGridBuilder` 으로 EstimatorParamMaps, `MulticlassClassificationEvaluator` 으로 Evaluator 를 만든다. 이를 통해 validator 를 다음과 같이 만들 수 있다.

```
val validator = new TrainValidationSplit().
  setSeed(Random.nextLong()).
  setEstimator(pipeline).
  setEvaluator(multiclassEval).
  setEstimatorParamMaps(paramGrid).
  setTrainRatio(0.9)
```

If the purpose of the CV set was to evaluate parameters that fit to the training set, then the purpose of the test set is to evaluate hyperparameters that were “fit” to the CV set. That is, the test set ensures an unbiased estimate of the accuracy of the final, chosen model and its hyperparameters.(CV set -> parameter 정하기, test set -> hyperparameter 정하기)

When a decision tree has overfit, it will exhibit high accuracy when run on the same training data that it fit the model to, but low accuracy on other examples. Here, the final model’s accuracy was about 91% on other, new examples. Accuracy can just as easily be evaluated over the same data that the model was trained on, trainData. This gives an accuracy of about 95%. The difference is not large but suggests that the decision tree has overfit the training data to some extent. A lower maximum depth might be a better choice.(train set 의 정확도가 test set 의 정확도보다 높다 -> overfitting)

### 4.10 범주형 특징 다시 살펴보기

현재까지는 `Wilderness_Area (4 binary columns)`, `Soil_Type (40 binary columns)` 를 numeric column 으로 생각해서 의사 결정 트리를 만들었으나 `qualitative - 0 (absence) or 1 (presence)` 의 특징을 가지는 categorical feature 이다.

Of course, this encoding forces the decision tree algorithm to consider the values of the underlying categorical features individually. Because features like soil type are broken down into many features, and because decision trees treat features individually, it is harder to relate information about related soil types.(하나의 특징을 여러 column 에 나눠 저장했기 때문에 별개의 것으로 생각해서 생기는 문제가 있을 수 있다.)

However, having 40 numeric features represent one 40-valued categorical feature increases memory usage and slows things down. This would replace, for example, the four columns encoding wilderness type with one column that encodes the wilderness type as a number between 0 and 3, like "Cover_Type".(메모리 사용이 많고 속도가 느려짐 -> 그냥 0 ~ 3, 0 ~ 39 로 표현하자)

VectorIndexer 의 역할 (소스 코드 주석)
- Automatically identify categorical features (default behavior)
  - This helps process a dataset of unknown vectors into a dataset with some continuous features and some categorical features. The choice between continuous and categorical is based upon a maxCategories parameter.
  - Set maxCategories to the maximum number of categorical any categorical feature should have.
  - E.g.: Feature 0 has unique values {-1.0, 0.0}, and feature 1 values {1.0, 3.0, 5.0}. If maxCategories = 2, then feature 0 will be declared categorical and use indices {0, 1}, and feature 1 will be declared continuous.

The approach assumes that the training set contains all possible values of each of the categorical features at least once. That is, it works correctly only if all 4 soil values and all 40 wilderness values appear in the training set so that all possible values get a mapping.(VectorIndexer 를 사용하기 위해서는 MaxCategories 로 설정한 수만큼 모든 값이 등장해야 함)

-> test set 에 대해 0.929 의 정확도, train set 에 대해 0.972 의 정확도가 나왔다.

### 4.11 랜덤 포레스트

Instead, decision trees use several heuristics to determine which few rules to actually consider. The process of picking rules also involves some randomness; only a few features picked at random are looked at each time, and only values from a random subset of the training data. This trades a bit of accuracy for a lot of speed, but it also means that the decision tree algorithm won’t build the same tree every time. This is a good thing.(모든 경우를 다 해볼 수는 없음 -> 매번 달라짐)

It would be great to have not one tree, but many trees, each producing reasonable but different and independent estimations of the right target value. Their collective average prediction should fall close to the true answer, more than any individual tree’s does. It’s the randomness in the process of building that helps create this independence. This is the key to random decision forests. The prediction of a random decision forest is simply a weighted average of the trees’ predictions. For a categorical target, this can be a majority vote or the most probable value based on the average of probabilities produced by the trees.(여러 랜덤 결정들의 평균이 실제에 가까워짐 -> random forest 원리)

Note that this classifier has another hyperparameter: the number of trees to build. Like the max bins hyperparameter, higher values should give better results up to a point. The cost, however, is that building many trees of course takes many times longer than building one.(the number of trees -> 새로운 hyperparamter)

Random decision forests are appealing in the context of big data because trees are supposed to be built independently, and big data technologies like Spark and Map‐Reduce inherently need data-parallel problems, where parts of the overall solution can be computed independently on parts of the data.(data-parallel 처리가 가능한 문제라서 매력적임)

당연하게도 # of trees 를 10 으로 했을 때가 최적 모델이었다 -> test set 에 대해 0.945, train set 에 대해 0.978

### 4.12 예측하기

The resulting "best model" is actually a whole pipeline of operations, which encapsulate how input is transformed for use with the model and includes the model itself, which can make predictions.(실제 데이터를 훈련시킨 모델에 그대로 넣을 수 있음)

### 4.13 한 걸음 더 나아가기

- Naïve Bayes
- Gradient boosting
- Logistic regression
- Multilayer perceptron

Each of these algorithms operates quite differently from decision trees and forests. However, many elements are the same: they plug into a Pipeline and operate on columns in a data frame, and have hyperparameters that you must select using training, cross-validation, and test subsets of the input data.(다양한 다른 알고리즘들도 존재하지만 과정은 비슷)
